pkg/controllers/nodepool/readiness/controller.go

```diff
@@ -0,0 +1,120 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package readiness
+
+import (
+	"context"
+
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	logger "log"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/awslabs/operatorpkg/status"
+
+	"sigs.k8s.io/karpenter/pkg/utils/nodepool"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"sigs.k8s.io/karpenter/pkg/apis/v1beta1"
+)
+
+// Controller for the resource
+type Controller struct {
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+}
+
+// NewController is a constructor
+func NewController(kubeClient client.Client, cloudProvider cloudprovider.CloudProvider) *Controller {
+	return &Controller{
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+	}
+}
+
+func (c *Controller) Reconcile(ctx context.Context, nodePool *v1beta1.NodePool) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, "nodepool.status")
+	stored := nodePool.DeepCopy()
+	supportedNC := c.cloudProvider.GetSupportedNodeClasses()
+
+	if len(supportedNC) == 0 {
+		logger.Fatal("no supported node classes found for the cloud provider")
+	}
+	nodeClass, err := c.getNodeClass(ctx, nodePool, supportedNC)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return reconcile.Result{}, err
+		}
+		log.FromContext(ctx).WithValues("NodeClass", nodePool.Spec.Template.Spec.NodeClassRef.Name).Info("nodeClass not found")
+		nodePool.StatusConditions().SetFalse(v1beta1.ConditionTypeNodeClassReady, "UnresolvedNodeClass", "Unable to resolve nodeClass")
+	} else {
+		c.setReadyCondition(nodePool, nodeClass)
+	}
+	if !equality.Semantic.DeepEqual(stored, nodePool) {
+		if err = c.kubeClient.Status().Update(ctx, nodePool); client.IgnoreNotFound(err) != nil {
+			if errors.IsConflict(err) {
+				return reconcile.Result{Requeue: true}, nil
+			}
+			return reconcile.Result{}, err
+		}
+	}
+	return reconcile.Result{}, nil
+}
+func (c *Controller) getNodeClass(ctx context.Context, nodePool *v1beta1.NodePool, supportedNC []status.Object) (status.Object, error) {
+	nodeClass := supportedNC[0]
+	if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: nodePool.Spec.Template.Spec.NodeClassRef.Name}, nodeClass); err != nil {
+		return nil, err
```

Comment:
you could do the ignore not found error here, and then when you consume the results of this you can just check with 

```
if err != nil {
    return
} 
if nodeclass != nil {
    setNotReady()
} else {
    setReady()
}
```

Personally, i think that's cleaner, but up to you

Reply:
Realized that nodeClass will never will be nil in this case, that's because cloudProvider returns []status.Object which is populates with nodeClass object. For example EC2NodeClass in case of aws provider.

Reply:
If the NodeClass doesn't exist, we should return it as nil, no?

Reply:
If we check for `notFounderror` like this, then we don't return nil. This is how I had it earlier -
```
if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: nodePool.Spec.Template.Spec.NodeClassRef.Name}, nodeClass); client.ignoreNotFound(err)!=nil {
      return nil, err
}
return nodeClass,nil
```
I have updated it now so that it checks for notFound and then returns nil.